core/store/src/trie/state_snapshot.rs

```diff
@@ -174,126 +173,89 @@ impl ShardTries {
             tracing::info_span!(target: "state_snapshot", "make_state_snapshot", ?prev_block_hash)
                 .entered();
         tracing::info!(target: "state_snapshot", ?prev_block_hash, "make_state_snapshot");
-        match &self.state_snapshot_config() {
-            StateSnapshotConfig::Disabled => {
-                tracing::info!(target: "state_snapshot", "State Snapshots are disabled");
-                Ok(())
+
+        let StateSnapshotConfig {
+            enabled, home_dir, hot_store_path, state_snapshot_subdir, ..
+        } = self.state_snapshot_config();
+
+        if !enabled {
+            tracing::info!(target: "state_snapshot", "State Snapshots are disabled");
+            return Ok(());
+        }
+
+        let _timer = metrics::MAKE_STATE_SNAPSHOT_ELAPSED.start_timer();
+        // `write()` lock is held for the whole duration of this function.
+        // Accessing the snapshot in other parts of the system will fail.
+        let mut state_snapshot_lock = self
+            .state_snapshot()
+            .write()
+            .map_err(|_| anyhow::Error::msg("error accessing write lock of state_snapshot"))?;
```

Comment:
nit: Can you embed the original error in the mapped error?

Reply:
Sure! This was part of a pure refactor tho...